src/executioners/OptimizeSolve.C

```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
 }
 
 PetscErrorCode
-OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat hessian, Mat /*pc*/, void * ctx)
+OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat /*hessian*/, Mat /*pc*/, void * ctx)
 {
-  auto * solver = static_cast<OptimizeSolve *>(ctx);
+  // this will need to set the real parameters (Vec x) on the solver (see ln 250) so that the
+  // current parameter can be used in the applyHessianWrapper.  This is not important for force
+  // inversion because the hessian does not change.  This will be important for material
+  // inversion.
 
+  auto * solver = static_cast<OptimizeSolve *>(ctx);
   libMesh::PetscVector<Number> param(x, solver->_my_comm);
   *solver->_parameters = param;
+  auto n = solver->_ndof;
+  auto comm = solver->_my_comm.get();
+  MatCreateShell(
+      comm, n, n, n, n, ctx, &(solver->_hessian)); // need to fix two of the n's for parallel
+  MatShellSetOperation(
+      solver->_hessian, MATOP_MULT, (void (*)(void))OptimizeSolve::applyHessianWrapper);
+  return 0;
+}
+
+PetscErrorCode
+OptimizeSolve::applyHessianWrapper(Mat H, Vec s, Vec Hs)
+{
+  void * ctx;
+  MatShellGetContext(H, &ctx);
+  auto * solver = static_cast<OptimizeSolve *>(ctx);
+  libMesh::PetscVector<Number> sbar(s, solver->_my_comm);
+  libMesh::PetscVector<Number> Hsbar(Hs, solver->_my_comm);
+  return solver->applyHessian(sbar, Hsbar);
+}
+
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> & s, libMesh::PetscVector<Number> & Hs)
+{
 
-  libMesh::PetscMatrix<Number> mat(hessian, solver->_my_comm);
+  //  if (!_problem.hasMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+  //    mooseError("Hessian based optimization algorithms require a sub-app with:\n"
+  //               "   execute_on = HOMOGENEOUS_FORWARD");
 
-  solver->hessianFunction(mat);
+  // What happens for material inversion when the Hessian
+  // is dependent on the parameters? Deal with it later???
+  // see notes on how this needs to change for Material inversion
+  _form_function->updateParameters(s);
+  if (!_problem.execMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+    mooseError("Forward solve multiapp failed!");
```

Comment:
I removed these errors because the test objects will fail because they don't have a subapp.  Maybe we need to do something about this interface in OptimizeSolve.C. Like can we derive something from it for when we have subapps.... I don't know if that will work since tao needs to be able to call static functions, I don't think those can be polymorphic.  Or we need to redesign the test objects to test the code in the way we are using it. I would like the test objects to provide a simple example of using tao directly.

Reply:
Done.